model/http-common.yaml

```diff
@@ -52,6 +52,13 @@ groups:
         requirement_level: required
         brief: >
           Port identifier of the ["URI origin"](https://www.rfc-editor.org/rfc/rfc9110.html#name-uri-origin) HTTP request is sent to.
+      - ref: http.route
+        requirement_level: opt_in
```

Comment:
I'm not sure if this is the right requirement level on the span side, e.g. I think this requirement level would prevent us from capturing it by default when we are able to, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10506

Reply:
@trask would changing it to `conditionally required` (when the instrumentation is able to capture route), be still OK when it comes to breaking changes? The opt_in was added to ensure we don't introduce a breaking change.

Reply:
it needs to be opt-in on existing metrics (#722), but it doesn't need to be opt-in on spans

a good question is whether we want it to be consistent (opt-in for both)

I think I'd personally suggest that it be recommended when available on spans (and maybe even influenced the span name?)

but would like to hear your and others' thoughts on that (@lmolkova @jsuereth @pyohannes @joaopgrassi @alanwest who all commented on #722)

Reply:
> and maybe even influenced the span name?

oh wait probably changes to the span name would be breaking...

Reply:
given it's opt-in I'm not sure it's breaking (as long as we document it).  It also brings substantial improvements to HTTP span names without taking away any old functionality.

Someone who wanted to query/group based on the old client name (HTTP method), but opts into route, can update their queries to use method alone (if that's what they want).

But probably they wanted something else - like get N longest operations based on name - and this case would just get better.

Reply:
> it needs to be opt-in on existing metrics (https://github.com/open-telemetry/semantic-conventions/issues/722), but it doesn't need to be opt-in on spans

great question. 
Yes, if it's not opt-in then changing span name is technically breaking. But I still think it's a good change and struggle to come up with the examples when someone would not want a route on the span name

Reply:
so, opt-in covering all 3?
* span attribute `http.route` on HTTP client spans
* metric attribute `http.route` on `http.client.request.duration`
* better span name on HTTP client spans

what do you think if a distro chooses to enable these by default in a major version bump (e.g. OpenTelemetry Java agent 3.0)? maybe we need a new requirement level for things that are only opt-in due to backwards compatibility, to allow instrumentations to enable them in a major version bump?

Reply:
> what do you think if a distro chooses to enable these by default in a major version bump (e.g. OpenTelemetry Java agent 3.0)?

Strictly speaking, that wouldn't be in line with [how the opt-in requirement level is defined](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/#opt-in):

> Instrumentations SHOULD populate the attribute if and only if the user configures the instrumentation to do so. Instrumentation that doesn’t support configuration MUST NOT populate Opt-In attributes.

The main problem is that we abuse the opt-in requirement level for use cases it's not intended for.

That's what the specification says:

> This attribute requirement level is recommended for attributes that are particularly expensive to retrieve or might pose a security or privacy risk.

We on the other hand use it to work around stability and backward compatibility requirements. I'm not sure about introducing another requirement level just for that, it seems somehow broken.

An alternative might be to define additional attributes with a requirement level that makes sense (e. g. "Recommended"), but mark them as experimental (or define them in an experimental extension), and thus put the changes into a queue for upcoming breaking changes for semantic conventions. Honestly, given the scope, dynamism, and strict stability and backward compatibility rules in semantic conventions, I think we should consider the case of major version increases which would allow us to make breaking changes (very rarely though). [This happened for ECS](https://github.com/elastic/ecs/blob/main/CHANGELOG.md#breaking-changes), and I wonder about their related experiences.

Reply:
@trask @klauco What do you think about adding this as an experimental extension to existing stable conventions (for now)? That would make it easier to get things into place for prototyping and experiments.

It seems rather exceptional to me that we merge something as stable that hasn't been in an experimental state before.

Reply:
> how do you envision marking http.route (stable) as experimental?

I don't have a good proposal here, there might be an section (marked as experimental) which lists  additional (stable or experimental) attributes. This seems somehow related to https://github.com/open-telemetry/semantic-conventions/issues/906.

> so I don't see any contradiction with how we use it, but it could be good idea to clarify.

You're right, we're not violating the specification. However, it seems to me we're using "Opt-In" for use cases it wasn't originally intended for.

Reply:
@trask 
> what do you think if a distro chooses to enable these by default in a major version bump (e.g. OpenTelemetry Java agent 3.0)? 

If I use just my judgement, behavior/runtime breaking changes are somewhat acceptable for most of the (even very conservative) libraries/tools with minor version bump. I look at .NET or Azure SDK changelogs. 

If we pick `http.route` as the best option, then not allowing users to use it in my opinion does more harm then good.
I still find it hard to imagine someone building alert based on HTTP client span name (like `GET`) and this alert being reliable.

Given all of this, based on my own judgement, I'd vote for making `http.route` (or a new attribute) `recommended: if available` and using it as a span name even without major version bump.

If we leave my opinion aside, we need a framework/process to decide what we do with such breaking changes - if we don't allow them, we'd not be able to move forward.

Reply:
> If we leave my opinion aside, we need a framework/process to decide what we do with such breaking changes on the semconv level - if we don't allow them, we'd not be able to move forward and fix old mistakes.

I fully agree. Discussing enabling opt-in attributes by default shows a certain danger of reality not aligning with what we specify in semantic conventions, which puts some of the work we're doing here into question.

Reply:
Could we resolve this discussion since the PR now adds the attribute with opt-in requirement level and we've captured other follow up steps in the #923 ?